libkbfs/folder_branch_ops.go

```diff
@@ -2494,16 +2500,80 @@ func (fbo *folderBranchOps) createEntryLocked(
 		newBlock = &FileBlock{}
 	}
 
-	de, err := fbo.syncBlockAndFinalizeLocked(
-		ctx, lState, md, newBlock, dirPath, name, entryType,
-		true, true, zeroPtr, excl)
+	// Cache update and operations until batch happens.  Make a new
+	// temporary ID and directory entry.
+	newID, err := fbo.config.cryptoPure().MakeTemporaryBlockID()
+	if err != nil {
+		return nil, DirEntry{}, err
+	}
+	session, err := fbo.config.KBPKI().GetCurrentSession(ctx)
 	if err != nil {
 		return nil, DirEntry{}, err
 	}
-	node, err := fbo.nodeCache.GetOrCreate(de.BlockPointer, name, dir)
+
+	newPtr := BlockPointer{
+		ID:         newID,
+		KeyGen:     md.LatestKeyGeneration(),
+		DataVer:    fbo.config.DataVersion(),
+		DirectType: DirectBlock,
+		Context: kbfsblock.MakeFirstContext(
+			session.UID, keybase1.BlockType_DATA),
+	}
+	node, err := fbo.nodeCache.GetOrCreate(newPtr, name, dir)
 	if err != nil {
 		return nil, DirEntry{}, err
 	}
+	de := DirEntry{
+		BlockInfo: BlockInfo{
+			BlockPointer: newPtr,
+			EncodedSize:  0,
+		},
+		EntryInfo: EntryInfo{
+			Type: entryType,
+			Size: 0,
+		},
+	}
+	fbo.blocks.AddDirEntryInCache(lState, dirPath, name, de)
+	fbo.dirOps = append(fbo.dirOps, cachedDirOp{co, []Node{dir, node}})
+
+	cleanupFn := func() {
+		// Until KBFS-2076 is done, clear out the cached dir data manually.
+		fbo.dirOps = nil
+		fbo.blocks.ClearCachedAddsAndRemoves(lState, dirPath)
+		fbo.blocks.ClearCacheInfo(lState, dirPath.ChildPath(name, newPtr))
+	}
+	defer cleanupFn()
+
+	de, err = fbo.syncBlockAndFinalizeLocked(
+		ctx, lState, md, newBlock, dirPath, name, entryType,
+		true, true, zeroPtr, excl)
+	if excl == WithExcl && errors.Cause(err) == errNoUpdatesWhileDirty {
+		// If an exclusive write hits a conflict, it will try to
+		// update, but won't be able to because of the dirty directory
+		// entries.  We need to clean up the dirty entries here first
+		// before trying to apply the updates again.
+		fbo.log.CDebugf(ctx, "Clearing dirty entries before applying new "+
+			"updates for exclusive write")
+		cleanupFn()
+		err = fbo.getAndApplyMDUpdates(ctx, lState, fbo.applyMDUpdatesLocked)
```

Comment:
This part is pretty ugly but I'm not sure how else to clean out the dirty data to allow the updates to be applied.